src/main/java/org/craftercms/engine/websocket/WsUpgradeHandler.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2007-2023 Crafter Software Corporation. All Rights Reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as published by
+ * the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.craftercms.engine.websocket;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpUpgradeHandler;
+import javax.servlet.http.WebConnection;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/**
+ * Upgrade websocket handler
+ */
+public class WsUpgradeHandler implements HttpUpgradeHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(WsUpgradeHandler.class);
+
+    ExecutorService exec;
+    InputStream socketIn;
+    OutputStream socketOut;
+    Socket socket;
+    Future<?> future;
+
+    public void preInit(ExecutorService exec, InputStream socketIn, OutputStream socketOut, Socket socket) {
+        this.exec = exec;
+        this.socketIn = socketIn;
+        this.socketOut = socketOut;
+        this.socket = socket;
+    }
+
+    /**
+     * Forward back and forth the websocket stream between websocket server and web client via engine as the proxy
+     * @param wc the WebConnection object associated to this upgrade request
+     */
+    @Override
+    public void init(WebConnection wc) {
+        logger.debug("* Websocket| Upgrade begin");
+        try {
+            var servletIn = wc.getInputStream();
```

Comment:
Similar concerns about memory and performance.

Reply:
Here we also do the same as the current HTTP Proxy is doing to get the stream and then read/write buffer to it: https://github.com/mitre/HTTP-Proxy-Servlet/blob/master/src/main/java/org/mitre/dsmiley/httpproxy/ProxyServlet.java#L676

Reply:
HTTPd/Tomcat/etc will limit the HTTP Headers size.